### ISSUE TITLE
CmdlineEnter Statusline Update

### ISSUE THREAD

**Issue by dylnmc** [vim/vim#4131]
TL;DR: when one presses <kbd>:</kbd> to enter command-mode, the statusline does not update.

Here is the conundrum. `mode(1)` reports the correct value of `'c'`. So, it is easy to check if the mode() is "command-mode". However, if one tries to get the statusline to update with, for example, `autocmd CmdlineEnter * redrawstatus` then it clears any ouput from a previous (series of) command(s). This means if one types `:ls` then presses <kbd>:</kbd> again to switch to a buffer, then the results of the `ls` disappear if `redrawstatus` command is called.

I feel it is a reasonable request (indeed neovim seems to have fixed it) to update the statusline on CmdlineEnter without redrawing everything. Is this possible?

~dylnmc

**Comment by chrisbra** [comment]
So something like this should fix it, right?

```patch
diff --git a/src/ex_getln.c b/src/ex_getln.c
index d40529d8c..796d3adf6 100644
--- a/src/ex_getln.c
+++ b/src/ex_getln.c
@@ -1014,6 +1014,14 @@ getcmdline_int(
     do_digraph(-1);            /* init digraph typeahead */
 #endif

+    // Redraw the statusline, if no message has been shown and redrawing is not disabled
+    // this allows to update the statusline that contains a mode() call
+    if (KeyTyped && !msg_scrolled && !p_lz && !RedrawingDisabled)
+    {
+       status_redraw_curbuf();
+       update_screen(0);
+    }
+
     /* If something above caused an error, reset the flags, we do want to type
      * and execute commands. Display may be messed up a bit. */
     if (did_emsg)
```

**Comment by dylnmc** [comment]
@chrisbra interesting. Yes; that will fix it when there are no messages. If there is no way to simply redraw the statusline by itself -- which would allow this patch to update the stl in other cases as well -- then this should work fine.

Thanks for the patch.

**Comment by zeertzjq** [comment]
I think this has been fixed by patch 8.2.2686 and 8.2.2737.
